🚀  [amp-story-player] Wait for first story to load before loading next ones

Enriqe opened the pull request:
Fixes #29320

Loading 2 iframes initially involved a lot of changes (because of how other methods like `show()` expect there to be 3 iframes) so I'm sending those in a follow up. (Tracked  #29706)

As seen on the first table, having a story specific event that delays the subsequent stories from being loaded makes the initial load time of the current story shorter.

### Time for first story page to be loaded 


|          | current | `storyLoaded`      | `documentLoaded`  |
|----------|---------|------------------|-----------------|
| normal   | 1.88s   | 1.37s (Δ -27.13%) | 1.74s (Δ -7.45%) |
| fast 3G  | ~72s    | ~30s (Δ -58.33%)  | ~34s (Δ -52.78%) |

There is a downside though, loading subsequent stories take longer to finish pre-rendering. 

### Time for secondary stories to be pre-rendered
|          | current | new             |
|----------|---------|-----------------|
| normal   | 3.17s   | 4.14s (Δ +30.6%) |
| fast 3G  | ~77s    | ~97s (Δ +25.97%) |

_Quick calculations using different network connection emulations and with the same player sample page containing local stories._

gmajoulet commented:
I filed this ticket for the followup to make it easier to track: https://github.com/ampproject/amphtml/issues/29706

gmajoulet commented on extensions/amp-story/1.0/amp-story.js:
```diff
@@ -1206,6 +1206,8 @@ export class AmpStory extends AMP.BaseElement {
       {bubbles: true}
     );
+    this.viewerMessagingHandler_ &&
+      this.viewerMessagingHandler_.send(EventType.STORY_LOADED, dict({}));
```
We need to find a name that's consistent with all the other AMP viewer events, and should consider getting it approved by the rest of the AMP team. There's already a `documentLoaded` message so I'm not sure what to pick here.
All the events we have for now could theoretically be re-used by regular AMP documents.

gmajoulet commented on src/amp-story-player/amp-story-player-impl.js:
```diff
@@ -551,20 +551,38 @@ export class AmpStoryPlayer {
+        firstStoryLoadedPromise = this.waitForStoryToLoadPromise_(iframeIdx);
+      } else {
+        // Only layout next story after first one is loaded.
+        firstStoryLoadedPromise.then(() => {
```
What happens if you swipe next before the first story is loaded?

Enriqe replied:
Good point, addressed in latest commit!

Enriqe commented:
* Added some metrics/calculations in PR description that measures this change's loading time.
* Added functionality to handle edge case where user swipes before first story is loaded.

Re. event name: Considering that having an amp-story specific message is faster than using `onDocumentReady`, we should use that instead. We should probably change the name to stick with the existing event names: maybe `onStoryReady`?

Enriqe commented on src/amp-story-player/amp-story-player-impl.js:
```diff
@@ -800,24 +814,29 @@ export class AmpStoryPlayer {
+  updateCurrentIframe_(story) {
+    const iframeIdx = story[IFRAME_IDX];
+    const iframeEl = this.iframes_[iframeIdx];
+    this.layoutIframe_(story, iframeEl, VisibilityState.VISIBLE).then(() => {
```
This `layoutIframe_` is a preventive operation in case the user swiped to navigate before this new current story started its `layoutIframe_` (because it was waiting for the first story to load).

Enriqe commented:
Friendly ping

gmajoulet commented on extensions/amp-story/1.0/amp-story.js:
```diff
@@ -1185,6 +1185,8 @@ export class AmpStory extends AMP.BaseElement {
       {bubbles: true}
     );
+    this.viewerMessagingHandler_ &&
+      this.viewerMessagingHandler_.send('storyLoaded', dict({}));
```
Two suggestions on the name:

1. We could try to make it more generic and dispatch an event like `initialContentLoaded` but I have no idea how we'd make this work for regular AMP so not sure it's a good idea.

2. If we make it Story specific let's make it very specific, it'd be hard to tell the difference between `documentLoaded` and `storyLoaded`. Wdyt of something that'd be like `storyContentLoaded` or `storyInitialContentLoaded` or anything you like?

gmajoulet commented on src/amp-story-player/amp-story-player-impl.js:
```diff
@@ -565,6 +569,21 @@ export class AmpStoryPlayer {
+   * @private
+   */
+  waitForStoryToLoadPromise_(iframeIdx) {
+    return new Promise((resolve) => {
```
Did you decide to not implement a timeout because of the fallback you have where if the user navigates to the next story it still loads properly?

gmajoulet commented on src/amp-story-player/amp-story-player-impl.js:
```diff
@@ -870,13 +891,52 @@ export class AmpStoryPlayer {
-      const {href} = this.getEncodedLocation_(url, visibilityState);
-      iframe.setAttribute('src', href);
-      iframe.setAttribute('title', story.textContent.trim());
+    return this.hasSameHref_(story.href, iframe.src).then((hasBeenLaidOut) => {
```
This could get much easier if you'd keep the previous code to retrieve the cacheURL and comparing it to the iframe src right here.

ie:

```
this.maybeGetCacheUrl_(story.href).then((url) => {
  if (sanitizedUrlsAreEquals) return;
});
```

Then you wouldn't have to retrieve the cache URL a second time.

gmajoulet commented on src/amp-story-player/amp-story-player-impl.js:
```diff
@@ -870,13 +891,52 @@ export class AmpStoryPlayer {
+      let waitPromise;
+      if (visibilityState === VisibilityState.VISIBLE) {
+        waitPromise = Promise.resolve();
+        this.initialStoryLoaded_ = this.waitForStoryToLoadPromise_(
```
Would `this.currentStoryLoadPromise_` better describe this property?

gmajoulet commented on src/amp-story-player/amp-story-player-impl.js:
```diff
@@ -870,13 +891,52 @@ export class AmpStoryPlayer {
+        return Promise.resolve();
+      }
+
+      let waitPromise;
```
When reassigning the `waitPromise` (aka navigation), I think it'd be great to reject the previous `waitPromise`. This way we know for sure that the success callbacks you registered will never trigger.

It'd protect ourselves from the following scenario:

1. Wait for active story to load (it takes a while), promise is pending and waiting to set a `src` on the next iframe
2. User swipes next, which loads the new story
3. First story finally loads and the callback triggers, setting the iframe `src` again

It's fine in this scenario (as long as it doesn't reload the iframe) but if users start adding stories through the programmatic API it can get weird.

Enriqe commented:
PTAL

Enriqe commented:
Friendly ping

Enriqe commented:
@gmajoulet friendly ping

gmajoulet commented on src/amp-story-player/amp-story-player-impl.js:
```diff
@@ -169,6 +171,15 @@ export class AmpStoryPlayer {
     };
 
+    /** @private {?Promise} */
+    this.currentStoryLoadPromise_ = null;
```
Nit: I think this file will grow a lot, combining these 3 into 1 `currentStoryLoadDeferred_` might be worth it
